Fix wrong type checking in concat

pbourguignon opened the pull request:
While the documentation of ``pandas.DataFrame.concat`` specifies that
any mapping of ``Label`` to ``FrameOrSeries`` is acceptable as the first
argument, but the code actually checks for an instance of ``dict``.
This changeset fixes that restriction.

- [x] closes #32863 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

pbourguignon committed:
Fix wrong type checking in concat

While the documentation of ``pandas.DataFrame.concat`` specifies that
any mapping of ``Label`` to ``FrameOrSeries`` is acceptable as the first
argument, the actual test performed checks for an instance of ``dict``.
This changeset fixes that restriction.

simonjayhawkins commented on pandas/core/reshape/concat.py:
```diff
@@ -315,7 +315,7 @@ def __init__(
             )
 
-        if isinstance(objs, dict):
+        if isinstance(objs, Mapping):
```
can you use Mapping from collections.abc here.

jreback replied:
even better, use is_dict_like which is our canonical test for these things

pbourguignon replied:
> even better, use is_dict_like which is our canonical test for these things

According to the Python compatibility statement, Python 3.6+ is required, which has the collections abcs. I'm quite agnostic here, but using the collections.abc would feel more future-proof without introducing restrictions.

pbourguignon replied:
> can you use Mapping from collections.abc here.

They are actually the same class, but sure, makes sense.

simonjayhawkins reviewed (changes requested):
Thanks @pbourguignon for the PR. we are currently in the process of adding type annotations in order to pick up these inconsistencies.

Can you add type annotations for objs. Also the parameter description for objs starts with "If a dict is passed...", can you update the docs.

can you also add a test to prevent regressions.

jreback commented:
also pls add a test for this (in the OP)

pbourguignon committed:
Add test and fix docstring for pd.concat objs argument

This changeset adds a test asserting that pd.concat works as expected
with an `objs` argument of type `Mapping`. It also replaces the explicit
reference to the type `dict` in the corresponding entry of the
docstring.

pep8speaks commented:
Hello @pbourguignon! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/reshape/test_concat.py`](https://github.com/pandas-dev/pandas/blob/fd5aaa9df4c6e1232a5978971a8538398b7302e4/pandas/tests/reshape/test_concat.py):

> [Line 1262:9](https://github.com/pandas-dev/pandas/blob/fd5aaa9df4c6e1232a5978971a8538398b7302e4/pandas/tests/reshape/test_concat.py#L1262): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

pbourguignon committed:
Fix pep8 complaints

simonjayhawkins commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -1243,6 +1244,44 @@ def test_concat_dict(self):
         tm.assert_frame_equal(result, expected)
 
+    def test_concat_mapping(self):
+        class SomeMapping(Mapping):
```
thanks @pbourguignon for the update. we have a fixture for this...non_mapping_dict_subclass

pbourguignon replied:
> thanks @pbourguignon for the update. we have a fixture for this...non_mapping_dict_subclass

Sorry, missed that one. Will update accordingly.

pbourguignon replied:
> thanks @pbourguignon for the update. we have a fixture for this...non_mapping_dict_subclass

I suspect it was actually meant to be called "non_dict_mapping_subclass"... Should I change it and fix all occurrences? Better in a separate PR?

simonjayhawkins replied:
makes sense. normally a separate PR for unrelated changes, but since only a few occurrences could do it here.

pbourguignon replied:
> makes sense. normally a separate PR for unrelated changes, but since only a few occurrences could do it here.

For the sake of others' awareness, I'll prepare a separate PR for that once this one is merged in.

pbourguignon committed:
Use conftest defined fixture non_mapping_dict_subclass

pbourguignon committed:
Cleanup

pbourguignon commented:
One last question: should I add an entry to the existing whatsnew for 1.1.0?

pbourguignon committed:
Fix ordering of imports

simonjayhawkins commented:
yes

pbourguignon committed:
Add whatsnew entry for the bugfix.

simonjayhawkins commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -380,6 +380,7 @@ Reshaping
 - :meth:`DataFrame.replace` and :meth:`Series.replace` will raise a ``TypeError`` if ``to_replace`` is not an expected type. Previously the ``replace`` would fail silently (:issue:`18634`)
 - Bug in :meth:`DataFrame.apply` where callback was called with :class:`Series` parameter even though ``raw=True`` requested. (:issue:`32423`)
 - Bug in :meth:`DataFrame.pivot_table` losing timezone information when creating a :class:`MultiIndex` level from a column with timezone-aware dtype (:issue:`32558`)
+- Fixed incorrect type checking in ``pd.concat`` where annotations hinted to ``objs`` being of type ``Mapping``, but only an instance of ``dict`` would be accepted. (:issue:`32863`)
```
can you phrase this something like.. Bug in pd.concat when passing a non-dict mapping raised a TypeError

pbourguignon replied:
Done

pbourguignon committed:
Fix whatsnew entry

pbourguignon committed:
Merge branch 'master' into fix-type-assertion

simonjayhawkins commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -385,8 +385,10 @@ Reshaping
 - :meth:`DataFrame.replace` and :meth:`Series.replace` will raise a ``TypeError`` if ``to_replace`` is not an expected type. Previously the ``replace`` would fail silently (:issue:`18634`)
 - Bug in :meth:`DataFrame.apply` where callback was called with :class:`Series` parameter even though ``raw=True`` requested. (:issue:`32423`)
 - Bug in :meth:`DataFrame.pivot_table` losing timezone information when creating a :class:`MultiIndex` level from a column with timezone-aware dtype (:issue:`32558`)
+- Bug in ``pd.concat`` where when passing a non-dict mapping as ``objs`` would raise a TypeError. (:issue:`32863`)
```
```suggestion
- Bug in :meth:``pd.concat`` where when passing a non-dict mapping as ``objs`` would raise a TypeError (:issue:`32863`)
```

simonjayhawkins commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -1243,6 +1243,29 @@ def test_concat_dict(self):
         expected = concat([frames[k] for k in keys], keys=keys)
         tm.assert_frame_equal(result, expected)
 
+    def test_concat_mapping(self, non_mapping_dict_subclass):
```
can we reuse test_concat_dict using dict and non_mapping_dict_subclass for the constructor using pytest.mark.parametrize?

pbourguignon replied:
Let me check that.

simonjayhawkins replied:
does this work?

```python
    @pytest.mark.parametrize("mapping", ["mapping", "dict"])
    def test_concat_dict(self, mapping, non_mapping_dict_subclass):
        constructor = non_mapping_dict_subclass if mapping == "mapping" else dict
        frames = constructor({
            "foo": DataFrame(np.random.randn(4, 3)),
            "bar": DataFrame(np.random.randn(4, 3)),
            "baz": DataFrame(np.random.randn(4, 3)),
            "qux": DataFrame(np.random.randn(4, 3)),
        })
```

pbourguignon replied:
Neat workaround! Let me update the PR.

simonjayhawkins reviewed (commented):
Thanks @pbourguignon for working on this. couple more suggestions.

pbourguignon committed:
Update doc/source/whatsnew/v1.1.0.rst

Co-Authored-By: Simon Hawkins <simonjayhawkins@gmail.com>

pbourguignon committed:
Black reformatting

pbourguignon committed:
Merge remote-tracking branch 'origin/fix-type-assertion' into fix-type-assertion

pbourguignon committed:
Deduplicate tests for dict/mapping input to pd.concat using a pytest.parametrize construct

pbourguignon committed:
Apply black

simonjayhawkins commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -1243,6 +1243,39 @@ def test_concat_dict(self):
+            pytest.param("dict", id="Built-in dict"),
+        ],
+    )
+    def test_concat_mapping(self, mapping, non_mapping_dict_subclass):
```
you can remove test_concat_dict now?

simonjayhawkins commented on pandas/tests/reshape/test_concat.py:
```diff
@@ -1243,6 +1243,39 @@ def test_concat_dict(self):
+    @pytest.mark.parametrize(
+        "mapping",
+        [
+            pytest.param("mapping", id="Non dict subclass of collections.abc.Mapping"),
```
since the parameters are strings, the strings will be the test ids, so no need to add these.

pbourguignon committed:
Remove duplicate test

jreback commented:
closing in favor of #32953

jreback commented:
@pbourguignon you have 2 PR for the same issue;. I closed tihs one as the other is more recent (and has the changes here too). If this is not correct pls comments.

pbourguignon commented:
> @pbourguignon you have 2 PR for the same issue;. I closed tihs one as the other is more recent (and has the changes here too). If this is not correct pls comments.

That's alright, I just couldn't find how to create the later PR against the earlier one.
